AppCenter/AppCenter/Internals/Channel/MSChannelUnitDefault.m

```diff
@@ -280,32 +280,34 @@ - (void)sendLogArray:(NSArray<id<MSLog>> *__nonnull)logArray
-                                                     if (logArray.count > 0) {
-
-                                                       // We have data to send.
-                                                       [self sendLogArray:logArray withBatchId:batchId andAuthToken:tokenInfo];
-                                                     } else {
-
-                                                       // No logs available with given params.
-                                                       if (tokenIndex == 0 && tokenArray[tokenIndex].endTime != nil &&
-                                                           [self.storage countLogsBeforeDate:tokenArray[tokenIndex].endTime] == 0) {
-
-                                                         // Delete token from history if we don't have logs fitting it in DB.
-                                                         [[MSAuthTokenContext sharedInstance] removeAuthToken:tokenInfo.authToken];
-                                                       }
-                                                       if (tokenIndex + 1 < tokenArray.count) {
-
-                                                         // Iterate to next token in array.
-                                                         [self flushQueueForTokenArray:tokenArray withTokenIndex:tokenIndex + 1];
-                                                       }
-                                                     }
-                                                   }];
+  self.availableBatchFromStorage =
+      [self.storage loadLogsWithGroupId:self.configuration.groupId
+                                  limit:self.configuration.batchSizeLimit
+                     excludedTargetKeys:[self.pausedTargetKeys allObjects]
+                              afterDate:tokenInfo.startTime
+                             beforeDate:tokenInfo.endTime
+                      completionHandler:^(NSArray<id<MSLog>> *_Nonnull logArray, NSString *batchId) {
+                        if (logArray.count > 0) {
+
+                          // We have data to send.
+                          [self sendLogArray:logArray withBatchId:batchId andAuthToken:tokenInfo];
+                        } else {
+
+                          // No logs available with given params.
+                          if (tokenIndex == 0 && tokenArray[tokenIndex].endTime != nil &&
+                              [self.storage countLogsBeforeDate:tokenArray[tokenIndex].endTime] == 0) {
+
+                            // Delete token from history if we don't have logs fitting it in DB.
+                            [[MSAuthTokenContext sharedInstance] removeAuthToken:tokenInfo.authToken];
+                          }
+                          if (tokenIndex + 1 < tokenArray.count) {
+                            [[MSAuthTokenContext sharedInstance] checkIfTokenNeedsToBeRefreshed:tokenArray[tokenIndex]];
```

Comment:
This doesn't handle the first token that is also the last token. Also, it refreshes the token after logs are sent, it is too late to refresh.

Reply:
@jaelim-ms It's storage's `completionHandler`, not sending network operation completion

Reply:
It still feels suspicious.

Reply:
Yeah I see the problem now. Fixing

Reply:
As for the updating after the logs are sent - we only fetch logs within time bounds of token validity info, thus we are unable to send logs with incorrect token. What had to be fixed - is triggering sending logs after token is refreshed (just in case we have smth else). So, we can call `checkIfTokenNeedsToBeRefreshed` right at the beginning of the callback

Reply:
That solves the issue but introduces a new one, though minor and can be addressed in another PR:
The check will be performed twice for the same token if multiple batches sent for the same token.